packages/x-tree-view-pro/src/internals/plugins/useTreeViewLazyLoading/useTreeViewLazyLoading.ts

```diff
@@ -0,0 +1,296 @@
+      } finally {
+        // set loading state
+        instance.setTreeViewLoading(false);
+      }
+    },
+    [nestedDataManager, params.dataSource, cache, resetDataSourceState, store, instance],
+  );
+
+  const fetchItemChildren = React.useCallback(
+    async (id: TreeViewItemId) => {
+      const getChildrenCount = params.dataSource?.getChildrenCount || (() => 0);
+
+      const getTreeItems = params.dataSource?.getTreeItems;
+      if (!getTreeItems) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+
+      const parent = selectorItemMeta(store.value, id);
+      if (!parent) {
+        nestedDataManager.clearPendingRequest(id);
+        return;
+      }
+
+      const depth = parent.depth ? parent.depth + 1 : 1;
+
+      // handle loading here
+      instance.setDataSourceLoading(id, true);
+
+      // handle caching here
+      const cachedData = cache.get(id);
+
+      if (cachedData !== undefined && cachedData !== -1) {
+        nestedDataManager.setRequestSettled(id);
+        instance.addItems({ items: cachedData, depth, parentId: id, getChildrenCount });
+        instance.setDataSourceLoading(id, false);
+
+        return;
+      }
+      if (cachedData === -1) {
+        instance.removeChildren(id);
+      }
+
+      const existingError = selectorGetTreeItemError(store.value, id) ?? null;
+      if (existingError) {
+        instance.setDataSourceError(id, null);
+      }
+
+      try {
+        const getTreeItemsResponse = await getTreeItems(id);
+        nestedDataManager.setRequestSettled(id);
+
+        // set caching
+        cache.set(id, getTreeItemsResponse);
+        // update the items in the state
+        instance.addItems({ items: getTreeItemsResponse, depth, parentId: id, getChildrenCount });
+      } catch (error) {
+        const childrenFetchError = error as Error;
+        // handle errors here
+        instance.setDataSourceError(id, childrenFetchError);
+        instance.removeChildren(id);
+      } finally {
+        // unset loading
+        instance.setDataSourceLoading(id, false);
+
+        nestedDataManager.setRequestSettled(id);
+      }
+    },
+    [nestedDataManager, params.dataSource, instance, cache, store],
+  );
+
+  useInstanceEventHandler(instance, 'beforeItemToggleExpansion', async (eventParameters) => {
+    if (!isLazyLoadingEnabled) {
+      return;
+    }
+
+    if (selectorIsItemExpanded(store.value, eventParameters.itemId)) {
+      return;
+    }
+
+    eventParameters.isExpansionPrevented = true;
+    await instance.fetchItems([eventParameters.itemId]);
+    const fetchErrors = Boolean(selectorGetTreeItemError(store.value, eventParameters.itemId));
+    if (!fetchErrors) {
+      instance.setItemExpansion(eventParameters.event, eventParameters.itemId, true);
+      if (selectorIsItemSelected(store.value, eventParameters.itemId)) {
+        // make sure selection propagation works correctly
+        instance.selectItem({
```

Comment:
Did you check that it was not fring unwanted callback like `onSelectedItemsChange`?